### PR TITLE
Adds check to ensure weave removal

### PR DESCRIPTION
**Pull request by camilamacedo86**
#### What this PR does / why we need it:

It seems that weave has not been delete always:

```
Events:
  Type     Reason                  Age                   From               Message
  ----     ------                  ----                  ----               -------
  Normal   Scheduled               34m                   default-scheduler  Successfully assigned default/troubleshoot-copyfromhost-k2z6l-z8mbt to data-plane-6
  Warning  FailedCreatePodSandBox  34m                   kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "f844f630ec4a7ea399dd21239bd29297851f42bd336682b70b9e64ae720752b5": plugin type="weave-net" name="weave" failed (add): unable to allocate IP address: Post "http://127.0.0.1:6784/ip/f844f630ec4a7ea399dd21239bd29297851f42bd336682b70b9e64ae720752b5": dial tcp 127.0.0.1:6784: connect: connection refused
  Normal   SandboxChanged          33m (x3 over 34m)     kubelet            Pod sandbox changed, it will be killed and re-created.
  Warning  FailedKillPod           4m4s (x137 over 33m)  kubelet            error killing pod: failed to "KillPodSandbox" for "bc8a5255-e36b-49e6-ab70-dde2b4769a65" with KillPodSandboxError: "rpc error: code = Unknown desc = failed to destroy network for sandbox \"f844f630ec4a7ea399dd21239bd29297851f42bd336682b70b9e64ae720752b5\": plugin type=\"weave-net\" name=\"weave\" failed (delete): Delete \"http://127.0.0.1:6784/ip/f844f630ec4a7ea399dd21239bd29297851f42bd336682b70b9e64ae720752b5\": dial tcp 127.0.0.1:6784: connect: connection refused"
```

#### Which issue(s) this PR fixes:

Motivated by a support case.
Linked to [sc-77764]

#### Special notes for your reviewer:

We are here ONLY checking and ensuring that all kind of resources that should be deleted are properly deleted:

```
$ kubectl get daemonset,rolebinding,role,clusterrolebinding,clusterrole,serviceaccount,secret -n kube-system | grep weave | awk '{print $1}' 
daemonset.apps/weave-net
rolebinding.rbac.authorization.k8s.io/weave-net
role.rbac.authorization.k8s.io/weave-net
clusterrolebinding.rbac.authorization.k8s.io/weave-net
clusterrole.rbac.authorization.k8s.io/weave-net
serviceaccount/weave-net
secret/weave-passwd
```

## Steps to reproduce
We can verify this one by checking the logs of the testgrid. 

#### Does this PR introduce a user-facing change?

```release-note
Fixes issue where weave resources has not been deleting `: plugin type=\"weave-net\" name=\"weave\" failed (delete): Delete ` by checking and forcing when required. This bug fix was introduced for weave version 0.21.5 and future ones.  
```

### CODE REVIEW

File: addons/flannel/0.21.5/install.sh
Comment (emosbaugh) [inline addons/flannel/0.21.5/install.sh:286]: if you delete with force and grace-period=0 does kubectl wait for pods to be terminated and removed? if not can you log this with confidence?

File: addons/flannel/0.21.5/install.sh
Comment (emosbaugh) [inline addons/flannel/0.21.5/install.sh:280]: why not just target the weave pods? im concerned that other pods may collide or if i named my node with the name weave then all daemonsets would have weave in the name

```suggestion
    weave_resources=$(kubectl -n kube-system get pods --no-headers -l name=weave-net)
```

File: addons/flannel/0.21.5/install.sh
Comment (emosbaugh) [inline addons/flannel/0.21.5/install.sh:284]:
```suggestion
        kubectl -n kube-system delete pods -l name=weave-net --grace-period=180
```

File: addons/flannel/0.21.5/install.sh
Comment (emosbaugh) [inline addons/flannel/0.21.5/install.sh:290]:
```suggestion
            kubectl -n kube-system delete ds -l name=weave-net --grace-period=0 --force
```

File: addons/flannel/0.21.5/install.sh
Comment (emosbaugh) [inline addons/flannel/0.21.5/install.sh:287]:
```suggestion
        weave_resources=$(kubectl -n kube-system get pods --no-headers -l name=weave-net)
```

File: addons/flannel/0.21.5/install.sh
Comment (emosbaugh) [inline addons/flannel/0.21.5/install.sh:295]:
```suggestion
        weave_resources=$(kubectl -n kube-system get pods --no-headers -l name=weave-net)
```

File: addons/flannel/template/base/install.sh
Comment (emosbaugh) [inline addons/flannel/template/base/install.sh:280]: you must change all the same suggestions in this file as well

**Review by emosbaugh — CHANGES_REQUESTED** [review]

File: addons/flannel/template/base/install.sh
Comment (camilamacedo86) [inline addons/flannel/template/base/install.sh:280]:
```suggestion
    weave_resources=$(kubectl -n kube-system get pods --no-headers -l name=weave-net)
```

File: addons/flannel/0.21.5/install.sh
Comment (camilamacedo86) [inline addons/flannel/0.21.5/install.sh:284]: @emosbaugh I need came back it for as it was.
See that in your suggestion you changed it for we delete ONLY pods when we have other kinds of resource and we should remove ALL. Because of this we need :

> kubectl -n kube-system delete $(kubectl get all -n kube-system | grep weave | awk '{print $1}')

File: addons/flannel/0.21.5/install.sh
Comment (camilamacedo86) [reply addons/flannel/0.21.5/install.sh]: We cannot do that. See that we need to ensure that all kind of resources deleted above were removed and not only Pods. 
    

> kubectl -n kube-system delete daemonset weave-net
>     kubectl -n kube-system delete rolebinding weave-net
>     kubectl -n kube-system delete role weave-net
>     kubectl delete clusterrolebinding weave-net
>     kubectl delete clusterrole weave-net
>     kubectl -n kube-system delete serviceaccount weave-net
>     kubectl -n kube-system delete secret weave-passwd

File: addons/flannel/0.21.5/install.sh
Comment (camilamacedo86) [reply addons/flannel/0.21.5/install.sh]: We cannot do that. See that we need to ensure that all kind of resources deleted above were removed and not only Pods. 
    

> kubectl -n kube-system delete daemonset weave-net
>     kubectl -n kube-system delete rolebinding weave-net
>     kubectl -n kube-system delete role weave-net
>     kubectl delete clusterrolebinding weave-net
>     kubectl delete clusterrole weave-net
>     kubectl -n kube-system delete serviceaccount weave-net
>     kubectl -n kube-system delete secret weave-passwd

File: addons/flannel/0.21.5/install.sh
Comment (camilamacedo86) [inline addons/flannel/0.21.5/install.sh:283]:
```suggestion
        logWarn "Weave resources are still present. Deleting Weave resources with grace period up to 1 minute..."
```

File: addons/flannel/template/base/install.sh
Comment (camilamacedo86) [inline addons/flannel/template/base/install.sh:283]:
```suggestion
        logWarn "Weave resources are still present. Deleting Weave resources with grace period up to 1 minute..."
```

File: addons/flannel/0.21.5/install.sh
Comment (emosbaugh) [inline addons/flannel/0.21.5/install.sh:292]: do you really need a loop? wont that add much more time? also why mask the output? also you need `--ignore-not-found` otherwise it could fail if the resource is not already present which is not what you want

```suggestion
    if ! timeout 60 kubectl -n kube-system delete --ignore-not-found "${resources[@]}" --grace-period=60 ; then
        logWarn "Weave resources still present after the deletion attempt with a grace period of 1 minute."
    fi
```

**Review by emosbaugh — CHANGES_REQUESTED** [review]

File: addons/flannel/0.21.5/install.sh
Comment (emosbaugh) [reply addons/flannel/0.21.5/install.sh:292]: added some padding to the timeout as well

File: addons/flannel/0.21.5/install.sh
Comment (CpuID) [inline addons/flannel/0.21.5/install.sh:325]: we should add an extra `if` conditional here, to check if the `weave` device exists, and only run `ip link delete weave` if it doesn't exist.

had a scenario in https://github.com/replicated-collab/lightup-replicated/issues/85#issuecomment-1571344745 today where `ip link delete` will error out if it tries to delete a device that doesn't exist

doing this idempotently would be a good thing 👍

File: addons/flannel/template/base/install.sh
Comment (CpuID) [inline addons/flannel/template/base/install.sh:327]: see https://github.com/replicatedhq/kURL/pull/4555/files#diff-dd06f6bd0012cb6a930396f707affe1ce3a7ab8edd02145c46c2f334433fa11dR325 same thing here (idempotently run `ip link delete weave` only if the device exists)

**Review by CpuID — CHANGES_REQUESTED** [review]

File: addons/flannel/0.21.5/install.sh
Comment (camilamacedo86) [reply addons/flannel/0.21.5/install.sh:325]: @CpuID Lets to do it in another PR so that we make eiaser the review and we ensure that we are updating accordingly all places by once. https://github.com/replicatedhq/kURL/pull/4576

File: addons/flannel/template/base/install.sh
Comment (camilamacedo86) [reply addons/flannel/template/base/install.sh:327]: @CpuID Lets to do it in another PR so that we make eiaser the review and we ensure that we are updating accordingly all places by once. https://github.com/replicatedhq/kURL/pull/4576

File: addons/flannel/template/base/install.sh
Comment (emosbaugh) [inline addons/flannel/template/base/install.sh:358]: if you add these functions to the flannel add-on they must be prefixed with flannel_

```suggestion
function flannel_delete_resource_gracefully() {
```

File: addons/flannel/template/base/install.sh
Comment (emosbaugh) [inline addons/flannel/template/base/install.sh:365]:
```suggestion
function flannel_force_delete_resource() {
```

File: addons/flannel/template/base/install.sh
Comment (emosbaugh) [inline addons/flannel/template/base/install.sh:372]: there is already a function kubernetes_resource_exists

File: addons/flannel/template/base/install.sh
Comment (emosbaugh) [inline addons/flannel/template/base/install.sh:361]: i feel like suppressing output that you do not expect to fail is a bad choice

```suggestion
    timeout 70 kubectl -n "$namespace" delete "$resource" --ignore-not-found --grace-period=60 1>/dev/null
```

File: addons/flannel/template/base/install.sh
Comment (emosbaugh) [inline addons/flannel/template/base/install.sh:368]: i feel like suppressing output that you do not expect to fail is a bad choice

```suggestion
    timeout 60 kubectl -n "$namespace" delete "$resource" --ignore-not-found --force --grace-period=0 1>/dev/null
```

File: addons/flannel/template/base/install.sh
Comment (emosbaugh) [inline addons/flannel/template/base/install.sh:375]: on the other hand this output should be suppressed

File: addons/flannel/template/base/install.sh
Comment (emosbaugh) [inline addons/flannel/template/base/install.sh:318]: why loop here? why not delete all at once?

File: addons/flannel/template/base/install.sh
Comment (emosbaugh) [inline addons/flannel/template/base/install.sh:337]: why loop here? why not delete all at once?

**Review by emosbaugh — CHANGES_REQUESTED** [review]

File: addons/flannel/template/base/install.sh
Comment (camilamacedo86) [reply addons/flannel/template/base/install.sh:318]: I am not sure if old versions if kubectl can work fine with multiple delations within different kinds
Also, if that fails we will know in what resource that fails by looking at the logs

File: addons/flannel/0.21.5/install.sh
Comment (emosbaugh) [inline addons/flannel/0.21.5/install.sh:320]: so this could take up to 7*320 = 37 minutes to finish for all resources? that is unacceptable.

**Review by emosbaugh — CHANGES_REQUESTED** [review]

**Review by CpuID — APPROVED** [review]
overall this LGTM, and I feel like it's a step in the right direction 👍 

I'll defer to others on the kURL team to approve also before this should be merged, to be safe

File: addons/flannel/template/base/install.sh
Comment (emosbaugh) [inline addons/flannel/template/base/install.sh:289]: You cannot use a spinner function for a long running command as mentioned by @laverya

**Review by emosbaugh — CHANGES_REQUESTED** [review]

**Review by emosbaugh — APPROVED** [review]

**Comment by camilamacedo86** [comment]
Rook is the new test added and it has been falling.
It has an issue that must be fixed but it is out of the scope of this PR.

**Pull request merged**
